### ISSUE TITLE
FHC spamming errors in Browser Console fast enough to lock up Firefox for minutes at a time

### ISSUE THREAD

**Issue by ssokolow** [stephanmahieu/formhistorycontrol-2#36]
I noticed my Firefox's content process was consuming 50% of one core when the browser should be at idle, so I opened up the Browser Console to see what was going on.

While I'm not sure if Form History Control was responsible for the load, it quickly filled up the console with a repeating cycle of these errors, which caused the whole browser to seize up so badly that, after clicking the Close button on the Browser Console window, I had to wait over 15 minutes for it to acknowledge and for the browser to start responding again.

(I had a half-written report of a probably-spurious Browser Console error message from CanvasBlocker that I didn't want to lose.)

```
Error: Error: ID already exists: editfldMore contextmenu.js:322:5
Error: Error: ID already exists: editfldlastused contextmenu.js:322:5
Error: Error: ID already exists: editfld2695 contextmenu.js:322:5
Error: Error: ID already exists: editfld2689 contextmenu.js:322:5
Error: Error: ID already exists: editfld2687 contextmenu.js:322:5
Error: Error: ID already exists: editfld2686 contextmenu.js:322:5
Error: Error: ID already exists: editfld2685 contextmenu.js:322:5
Error: Error: ID already exists: editfld2684 contextmenu.js:322:5
Error: Error: ID already exists: editfld2682 contextmenu.js:322:5
Error: Error: ID already exists: editfld2681 contextmenu.js:322:5
Error: Error: ID already exists: editfld2680 contextmenu.js:322:5
Error: Error: ID already exists: editfld2679 contextmenu.js:322:5
Error: Error: ID already exists: editfldhostname contextmenu.js:322:5
Error: Error: ID already exists: editfld2695 contextmenu.js:322:5
Error: Error: ID already exists: editfld2618 contextmenu.js:322:5
Error: Error: ID already exists: editfld2603 contextmenu.js:322:5
Error: Error: ID already exists: editfld2584 contextmenu.js:322:5
Error: Error: ID already exists: editfld2583 contextmenu.js:322:5
Error: Error: ID already exists: editfld2582 contextmenu.js:322:5
Error: Error: ID already exists: editfld2581 contextmenu.js:322:5
Error: Error: ID already exists: editfld2580 contextmenu.js:322:5
Error: Error: ID already exists: editfld2579 contextmenu.js:322:5
Error: Error: ID already exists: editfld2578 contextmenu.js:322:5
Error: Error: ID already exists: editfld2695 contextmenu.js:322:5
Error: Error: ID already exists: editfld2618 contextmenu.js:322:5
Error: Error: ID already exists: editfld2603 contextmenu.js:322:5
Error: Error: ID already exists: editfld2584 contextmenu.js:322:5
Error: Error: ID already exists: editfld2583 contextmenu.js:322:5
Error: Error: ID already exists: editfld2582 contextmenu.js:322:5
Error: Error: ID already exists: editfld2581 contextmenu.js:322:5
Error: Error: ID already exists: editfld2580 contextmenu.js:322:5
Error: Error: ID already exists: editfld2579 contextmenu.js:322:5
Error: Error: ID already exists: editfld2578 contextmenu.js:322:5
Error: Error: ID already exists: editfld2666 contextmenu.js:322:5
Error: Error: ID already exists: editfld2665 contextmenu.js:322:5
Error: Error: ID already exists: editfld2661 contextmenu.js:322:5
Error: Error: ID already exists: editfld2660 contextmenu.js:322:5
Error: Error: ID already exists: editfld2659 contextmenu.js:322:5
Error: Error: ID already exists: editfld2658 contextmenu.js:322:5
Error: Error: ID already exists: editfld2657 contextmenu.js:322:5
Error: Error: ID already exists: editfld2656 contextmenu.js:322:5
Error: Error: ID already exists: editfld2655 contextmenu.js:322:5
Error: Error: ID already exists: editfld2654 contextmenu.js:322:5
Error: Error: ID already exists: editfld2689 contextmenu.js:322:5
Error: Error: ID already exists: editfld2687 contextmenu.js:322:5
Error: Error: ID already exists: editfld2686 contextmenu.js:322:5
Error: Error: ID already exists: editfld2685 contextmenu.js:322:5
Error: Error: ID already exists: editfld2684 contextmenu.js:322:5
Error: Error: ID already exists: editfld2682 contextmenu.js:322:5
Error: Error: ID already exists: editfld2681 contextmenu.js:322:5
Error: Error: ID already exists: editfld2680 contextmenu.js:322:5
Error: Error: ID already exists: editfld2679 contextmenu.js:322:5
Error: Error: ID already exists: editfld2678 contextmenu.js:322:5
Error: Error: ID already exists: editfld2689 contextmenu.js:322:5
Error: Error: ID already exists: editfld2687 contextmenu.js:322:5
Error: Error: ID already exists: editfld2686 contextmenu.js:322:5
Error: Error: ID already exists: editfld2685 contextmenu.js:322:5
Error: Error: ID already exists: editfld2684 contextmenu.js:322:5
Error: Error: ID already exists: editfld2682 contextmenu.js:322:5
Error: Error: ID already exists: editfld2681 contextmenu.js:322:5
Error: Error: ID already exists: editfld2680 contextmenu.js:322:5
Error: Error: ID already exists: editfld2679 contextmenu.js:322:5
Error: Error: ID already exists: editfld2678 contextmenu.js:322:5
Error: Error: ID already exists: editfld2695 contextmenu.js:322:5
Error: Error: ID already exists: editfld2689 contextmenu.js:322:5
Error: Error: ID already exists: editfld2687 contextmenu.js:322:5
Error: Error: ID already exists: editfld2686 contextmenu.js:322:5
Error: Error: ID already exists: editfld2685 contextmenu.js:322:5
Error: Error: ID already exists: editfld2684 contextmenu.js:322:5
Error: Error: ID already exists: editfld2682 contextmenu.js:322:5
Error: Error: ID already exists: editfld2681 contextmenu.js:322:5
Error: Error: ID already exists: editfld2680 contextmenu.js:322:5
Error: Error: ID already exists: editfld2679 contextmenu.js:322:5
```

I know it's Form History Control because, before it bogged down too far, I was able to check the tooltip and...

1. Form History Control is the only thing in either Firefox or any of my extensions which has a `background/contextmenu.js`
2. It says line 322 and `console.error(`Error: ${browser.runtime.lastError}`);` would explain that "Error: Error:" in the output.
3. Disabling Form History Control caused the problem to go away

## Environment

**OS:** Kubuntu Linux 14.04.5 LTS (64-bit)
**Browser:** Firefox Developer Edition 63.0b4 (64-bit) (Mozilla Build)
**Form History Control Version:** 2.2.0.0

My FHC settings are mostly the defaults, but I think I changed these:

1. Save a new version of a multiline field when last saved version is older than 5 minutes
2. ...or content length changed more than 200 characters
3. Remove form history not used in the past 30 days.

**Comment by stephanmahieu** [comment]
Hi, after some investigation I can not figure out what may have happened in this case. Have you experienced this behavior more often?
Any idea what happened just before, did you open a lot of tabs (or started a new session that opened a lot of tabs), or did you just removed some (or a lot of) formhistory entries?

**Comment by ssokolow** [comment]
> Any idea what happened just before, did you open a lot of tabs (or started a new session that opened a lot of tabs)

I have a session that I've been restoring since I first build my new Firefox Quantum profile which, at the time, had somewhere around 350 tabs spread across four windows (on three monitors), with the required system resources kept reasonable by the Auto Tab Discard extension. (At the moment, it has 405 tabs open.)

> or did you just removed some (or a lot of) formhistory entries?

I hadn't cleared any history entries. In fact, aside from configuring it initially, a couple of incidents months and months ago where FHC failed to save the text I wanted to recover, and checking the settings when filing this report, I hadn't yet needed to interact with it.

> Have you experienced this behavior more often?

When I discovered that it was the source of the problem, I shut off FHC and haven't turned it back on since. (I've been falling back on my old habit of composing long stuff in an external text editor and then copy-pasting it into the browser... or, failing that, copy-pasting into an external text editor before I submit.)

In case it helps, here's the portion of `about:support` relating to the extensions I have enabled right now:

```
Auto Tab Discard			0.2.9	true	{c2c003ee-bd69-42a2-b0e9-6f34222cb046}
Awesome RSS				1.3.4	true	{97d566da-42c5-4ef4-a03b-5a2e5f7cbcb2}
Behind The Overlay Revival	0.1.8	true	{c0e1baea-b4cb-4b62-97f0-278392ff8c37}
Bloody Vikings!			0.9.4	true	bloodyvikings@ffs.bplaced.net
CanvasBlocker				0.5.3	true	CanvasBlocker@kkapsner.de
Carefree Tumblr			0.1.0	true	{bb54be14-8b70-48a2-929b-2b26fa8bcca0}
Cookie AutoDelete			2.2.0	true	CookieAutoDelete@kennydo.com
cookies.txt				0.1		true	{12cf650b-1822-40aa-bff0-996df6948878}
Copy PlainText				1.8		true	copyplaintext@eros.man
Decentraleyes				2.0.8	true	jid1-BoFifL9Vbdl2zQ@jetpack
Disable HTML5 Autoplay		2017.12.20	true	disable-html5-autoplay@afnankhan
Download Star				2.0.2	true	{8cc0b007-e40b-46e8-9e50-e3bf021c94ab}
File Backups				0.3.5	true	file-backups@pmario.github.io
Google search link fix		1.6.7	true	jid0-XWJxt5VvCXkKzQK99PhZqAn7Xbg@jetpack
Greasemonkey			4.7		true	{e4a8a97b-f2ed-450b-b12d-ee082ba24781}
HistoryBlock				2.0		true	historyblock@kain
HTTPS Everywhere			2018.9.19	true	https-everywhere@eff.org
In-Page Pop-up Reporter	1.6		true	{feb799e2-29e2-4e35-b862-cc4e1842b6f5}
InlineDisposition Reloaded	2.0.1	true	{914d3a0b-6982-4e58-9a60-2e5d6cf6b09b}
Link Cleaner				1.5		true	{6d85dea2-0fb4-4de3-9f8c-264bce9a2296}
Multiple Tab Handler		2.1.4	true	multipletab@piro.sakura.ne.jp
Open in Sidebar			2018.7.26	true	open-link-in-sidebar@afnankhan
PinPatrol					0.4		true	@PinPatrol
Pure URL					1.3.0a	true	pure-url@jetpack
Random User-Agent		2.2.6	true	{b43b974b-1d3a-4232-b226-eaa2ac6ebb69}
Reddit Enhancement Suite	5.12.5	true	jid1-xUfzOsOFlzSOXg@jetpack
Reload in address bar		1.6		true	{e1ed7a80-7c11-4f7e-968b-79b551a0067f}
Resurrect Pages			7		true	{0c8fbd76-bdeb-4c52-9b24-d587ce7b9dc3}
Select Link Text			2017.3.26	true	select-link-text@afnankhan
Session Boss				1.5		true	sessionboss@william.wong
Stylus					1.4.22	true	{7a7a4a92-a2a0-41d1-9fd7-1e92480d612d}
SuperStop				0.3		true	superstop@gavinsharp.com
Tab Count				1.0		true	tabcount@grepper.net
Tree Style Tab				2.5.4	true	treestyletab@piro.sakura.ne.jp
uBlock Origin				1.16.20	true	uBlock0@raymondhill.net
uMatrix					1.3.14	true	uMatrix@raymondhill.net
Unpaywall				3.77		true	{f209234a-76f0-4735-9920-eb62507a54cd}
Web ScrapBook			0.27.0	true	webscrapbook@danny0838.addons.mozilla.org
```

Reddit Enhancement Suite wasn't installed when I reported this and I'll probably remove it again once I can find the time to whip up a userscript which just sets the old Reddit favicon on Reddit tabs without doing anything heavier.

**Comment by stephanmahieu** [comment]
I suspect that when all sessions are restored each corresponding browser-tab is activated in rapid succession triggering an action in the add-on that prepares the context menu for restoring editor fields which is a very cpu-intensive asynchronous operation. Having many open sessions/tabs will thus trigger an unnecessary heavy load. 
The fact that apparently each session becomes active (even for only a split second) when all sessions are restored is not something I anticipated. I have to investigate further to see if this is new behavior of the browser or maybe triggered by one of the other installed add-ons.

The way that it was supposed the work is that when you change tabs, the context menu is recalculated for the newly activated tab.

**Comment by stephanmahieu** [comment]
It has happened to me only once, a couple of months ago. Maybe it has something to do with a Firefox update or maybe it was a special feature from my virusscanner.
Does anyone experienced this massive session restoring more than once?

**Comment by ssokolow** [comment]
Unfortunately, I can't help you there.

I still haven't re-enabled FHC since I reported this because I consider it too unacceptable a risk.

I can however say that I have no virus scanner, so it couldn't have been that. (I'm a Linux user. No virus scanner is the normal state of things.)

**Comment by stephanmahieu** [comment]
Should be fixed with the 2.4.0.0 release
